src/js/components/Button/Badge.js

```diff
@@ -0,0 +1,161 @@
+    else if (dimension === 'height') {
+      if (pad.top)
+        pad = parseMetricToNum(theme.global.edgeSize[pad.top] || pad.top);
+      if (pad.bottom)
+        pad += parseMetricToNum(
+          theme.global.edgeSize[pad.bottom] || pad.bottom,
+        );
+      else if (pad.vertical) {
+        pad =
+          parseMetricToNum(
+            theme.global.edgeSize[pad.vertical] || pad.vertical,
+          ) * 2;
+      } else pad = 0;
+    } else if (dimension === 'width') {
+      if (pad.left)
+        pad = parseMetricToNum(theme.global.edgeSize[pad.left] || pad.left);
+      if (pad.right)
+        pad += parseMetricToNum(theme.global.edgeSize[pad.right] || pad.right);
+      else if (pad.horizontal) {
+        pad =
+          parseMetricToNum(
+            theme.global.edgeSize[pad.horizontal] || pad.horizontal,
+          ) * 2;
+      } else pad = 0;
+    }
+
+    // if content is tall/wide, let badge grow to fit. otherwise,
+    // make sure it's at least badge.size.medium dimensions
+    return `${Math.max(
+      Math.ceil(badgeContentRef.current.getBoundingClientRect()[dimension]) +
+        pad +
+        borderWidth,
+      parseMetricToNum(theme.button.badge.size.medium) + borderWidth,
+    )}px`;
+  }
+  return `${badgeContentRef.current.getBoundingClientRect()[dimension]}px`;
+};
+
+export const Badge = ({ children, content }) => {
+  const theme = useContext(ThemeContext);
+  const badgeContentRef = useRef();
+
+  const defaultBadgeDimension =
+    typeof content === 'boolean' ||
+    (content && content.value && typeof content.value === 'boolean')
+      ? // empty badge should be smaller. this value was chosen as a default
+        // after experimenting with various values
+        `${parseMetricToNum(theme.button.badge.size.medium) / 2}px`
+      : theme.button.badge.size.medium;
+
+  const [[width, height], setBadgeDimension] = useState([
+    defaultBadgeDimension,
+    defaultBadgeDimension,
+  ]);
+
+  // scale badge to fit its contents
+  // width informs how far to horizontally offset the badge
+  // height informs how far to vertically offset the badge
+  useEffect(() => {
+    if (
+      badgeContentRef &&
+      badgeContentRef.current &&
+      typeof content !== 'boolean'
+    ) {
+      setBadgeDimension([
+        getBadgeDimension('width', content, badgeContentRef, theme),
+        getBadgeDimension('height', content, badgeContentRef, theme),
+      ]);
+    }
+  }, [content, theme]);
+
+  // offset the badge so it overlaps content. when badge has content,
+  // offset should be 50%. when badge is empty, offset by a smaller amount to
+  // keep the badge closer to the content. this value was chosen as a
+  // reasonable default after testing with various grommet icons.
+  const divisor =
+    typeof content === 'boolean' || (content && content.value === true)
+      ? 3.5
+      : 2;
+  const offset = {
+    horizontal: `-${parseMetricToNum(width) / divisor}px`,
```

Comment:
You can shoot me if you like for this but ...

What about not adding `offset` to Stack and just doing the offsetting directly in the DOM via `element.style.`, calculating the offset distance via `useLayoutEffect()` within Badge.js?